sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -659,8 +659,27 @@ trait ParquetPartitionReaderBase extends Logging with Arm with ScanWithMetrics
             val readField = readDataSchema(writeAt)
             if (areNamesEquiv(clippedGroups, readAt, readField.name, isSchemaCaseSensitive)) {
               val origCol = table.getColumn(readAt)
-              val col = if (typeCastingNeeded && precisionList.nonEmpty) {
-                convertDecimal64ToDecimal32Wrapper(origCol, precisionList.dequeue())
+              val col: ColumnVector = if (typeCastingNeeded && precisionList.nonEmpty) {
+                ColumnUtil.ifTrueThenDeepConvertTypeAtoTypeB(origCol, readField.dataType,
```

Comment:
OK Two things.

1.  Why not just have a single API that is like `nestedBitCastAsNeeded(col: ColumnView, sparkType: DataType)` which would walk through the col and match up the structure with `sparkType` as it went and do a nested bit cast when needed.  Then you can manipulate `sparkType` ahead of time to look exactly how you want it to.  The fact that there is a `precisionList` and this code consumes it feels both kind of brittle and confusing to follow.
2. Why do we need this in the GpuParquetScan code at all?  This appears to be specific to the cache serializer. It has to be because we do *NOT* want this type of thing to happen anytime we are reading a normal parquet file. It is not necessarily something that needs to be fixed here. It just feels like ti would be much cleaner if we could read the parquet file normally and then have a stage that did the bit casts after it was already read in, and the bit casting happened in the serializer without leaking into GpuParquetScan itself.

Reply:
My idea what that you take the schema `StructType("a": DecimalType(10, -2), "b": IntegerType)` and manipulate it how you want. `StructType("a", IntegerType, "b" IntegerType)` then pass that into `nestedBitCastAsNeeded`. It would see that the first column "a" is a Decimal Type and know that it needs to be an IntegerType so it would do the bit cast.  The when going back after you read the data from parquet you just pass in the original schema into `nestedBitCastAsNeeded` and it would see that "a" is an integer type, but you want a `DecimalType(10, -2)` and do the bit cast.